### PR TITLE
feat: add stack.id

### PR DESCRIPTION
**Pull request by katcipis**
Adds stack.id configuration + the terramate.stack.id metadata (documented [here](https://github.com/mineiros-io/terramate/pull/439)).

### CODE REVIEW

File: hcl/hcl_stack_test.go [hunk 1]
```diff
@@ -0,0 +1,573 @@
...
+		"_id_again_",
+		"-id-again-",
+		"-id_mixed-",
+		"-id_numbers-0123456789-",
+		"maxsize_id_Test_should_Be_64_bytes_aNd_now_running_out_of_ID-aaa",
+	}
+	invalidIDs := []string{
+		"",
+		"*not+valid$",
+		"cachaça",
```
Comment (i4ki) [inline hcl/hcl_stack_test.go:41]: :joy:

File: hcl/hcl.go [hunk 2]
```diff
@@ -146,6 +155,31 @@ type TerramateParser struct {
...
+	if !stackIDRegex.MatchString(id) {
+		return StackID{}, errors.E("Stack ID %q doesn't match %q", id, stackIDRegex)
+	}
+	return StackID{id: &id}, nil
+}
+
+// Value returns the ID string value and true if this StackID is defined,
+// it returns "" and false otherwise.
+func (s StackID) Value() (string, bool) {
+	// This is like a poor's man option type.
```
Comment (i4ki) [inline hcl/hcl.go:173]: I dont think the comments are needed because it looks clear as `id` is a pointer. But it's your call, you can keep it.

Comment (katcipis) [reply hcl/hcl.go:173]: ah we can remove, not in love with the comment :laughing:

**Review by i4ki — APPROVED** [review]
lgtm but left some comments.

File: stack/stack.go [hunk 3]
```diff
@@ -162,6 +179,76 @@ func LookupParent(root, dir string) (S, bool, error) {
 	return l.lookupParentStack(dir)
 }
 
+// LoadAll loads all stacks inside the given rootdir.
+func LoadAll(rootdir string) ([]S, error) {
```
[the remark this answers is not in the document]
Comment (katcipis) [reply stack/stack.go:183]: I moved the logic here, terramate.ListStacks now call this: https://github.com/mineiros-io/terramate/pull/441/files#diff-71a171b5d5e0b997cd17d0c089fe66366ca924c35e229127f0b7ec0287e1e1efR23

Or I missed something ? (it seemed to me that this is stack package specific logic, both listing + checking for stack related invariants, etc).

[the remark this answers is not in the document]
Comment (katcipis) [reply stack/stack.go:183]: Nice catch nonetheless, don't let me copy shit around :laughing:

**Review by i4ki — APPROVED** [review]
:rocket:

**Review by i4ki — APPROVED** [review]

**Pull request merged**